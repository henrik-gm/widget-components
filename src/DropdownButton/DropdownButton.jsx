import React, { PropTypes, Component } from 'react';
import ReactDOM from 'react-dom';
import styles from './DropdownButton.scss';

class DropdownButton extends Component {

   constructor(props) {
      super(props);

      this.state = {
         selected: this.props.selected,
         dropDown: false
      };

      this.buttonElement = null;

      this.onBackgroundClick = this.onBackgroundClick.bind(this);
      this.onButtonClick = this.onButtonClick.bind(this);
   }

   // called on background click when drop down window is open
   onBackgroundClick(ev) {
      // setState prevents event from being processed by onOptionClick handler
      // so it is moved to the next cycle
      this.setState({ dropDown: false });
      window.document.documentElement.removeEventListener('click', this.onBackgroundClick, true);
      const btnIndex = ev.target.getAttribute('data-kw-dropdown-button-index');
      if (btnIndex !== null || this.state.selected === btnIndex) {
         this.setState({ selected: btnIndex });
         this.props.onChange(btnIndex);
      }
   }

   /**
    * Shows drop down box with available options.
    * @param {SyntheticEvent} event Click event
    */
   onButtonClick(event) {
      event.stopPropagation();
      this.setState({
         dropDown: true,
      });
      window.document.documentElement.addEventListener('click', this.onBackgroundClick, true); // add event to the capture phase instead of bubble phase
   }

   /**
    * Renders button.
    * @returns {XML}
    */
   render() {
      // calculating position of dropdown while in mobile mode
      let dropdownMobileStyling = {};
      if (this.state.dropDown && this.buttonElement) {
         const { top, bottom } = this.buttonElement.getBoundingClientRect();
         dropdownMobileStyling = {
            top: top,
            bottom: 'auto'
         };
         if (this.props.verticalAlignment === 'bottom') {
            dropdownMobileStyling = {
               top: 'auto',
               bottom: Math.max(document.documentElement.clientHeight, window.innerHeight || 0) - bottom
            };
         }
      }

      let cssHAlignClass = styles.leftAlign;
      if (this.props.horizontalAlignment === 'right') {
         cssHAlignClass = styles.rightAlign;
      }

      let cssVAlignClass = styles.topAlign;
      if (this.props.verticalAlignment === 'bottom') {
         cssVAlignClass = styles.bottomAlign;
      }

      return (
         <div className={styles.general}>
            <button
               ref={(node) => { this.buttonElement = node }}
               className={'KambiWidget-card-support-text-color ' + styles.button} onClick={this.onButtonClick}
            >
               {this.props.options[this.state.selected]}
               <i />
            </button>
            {
               this.state.dropDown &&
                  <ul
                     style={dropdownMobileStyling}
                     className={`${styles.dropDown}  ${cssHAlignClass} ${cssVAlignClass}`}
                  >
                     {this.props.options.map((option, i) => {
                        const classNames = [
                           'KambiWidget-card-background-color',
                           'KambiWidget-card-background-color--hoverable',
                           'KambiWidget-card-background-color--clickable'
                        ];

                        if (this.state.selected === i) {
                           classNames.push('KambiWidget-primary-color');
                        }

                        return (
                           <li
                              key={option}
                              data-kw-dropdown-button-index={i}
                              className={classNames.join(' ')}
                           >
                              {option}
                           </li>
                        );
                     })}
                  </ul>
            }
         </div>
      );
   }
}

DropdownButton.propTypes = {
   /**
    * Options array
    */
   options: PropTypes.arrayOf(PropTypes.string).isRequired,

   /**
    * Defines which option should be checked upon component creation
    */
   selected: PropTypes.number,

   /**
    * Option change handler, the function is passed the index that references the position of the element selected in the options array
    */
   onChange: PropTypes.func.isRequired,

   /**
    * Horizontal alignment of the dropdown box. If 'left' will match the left corner of the dropdown with the left corner of the button, if 'right' will match the right corner for the dropdown to the right corner of the button
    * If widget width is smaller than 925 the dropdown will ignore this option and will cover the full width of the widget
    */
   horizontalAlignment: PropTypes.oneOf(['left', 'right']),

   /**
    * Vertical alignment of the dropdown box. If 'top' will match the top corner of the dropdown with the top corner of the button, if 'bottom' will match the bottom corner for the dropdown to the bottom corner of the button
    */
   verticalAlignment: PropTypes.oneOf(['top', 'bottom'])
};

DropdownButton.defaultProps = {
   selected: 0,
   horizontalAlignment: 'right',
   verticalAlignment: 'top'
}

export default DropdownButton;
