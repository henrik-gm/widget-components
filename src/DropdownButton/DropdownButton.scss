.general {
   display: inline-block;
   position: relative;
}

.button {
   background: none;
   border: none;
   outline: none;
   padding: 7px 7px 7px 7px;
   position: relative;

   i {
      border-left: 5px solid transparent;
      border-right: 5px solid transparent;
      border-top: 6px solid;
      border-color-top: #999; // fallback in case of browser that does not support currentColor
      border-color-top: currentColor;
      display: inline-block;
      margin-left: 5px;
      margin-bottom: 1px;
   }
}

.dropDown {
   position: absolute;
   z-index: 1;
   list-style: none;
   margin: 4px;
   padding: 0;
   border-radius: 2px;
   box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.5);
   min-width: 280px;
   text-align: left;

   li {
      font-size: 16px;
      padding: 17px 16px;
      line-height: 20px;
      cursor: pointer;
      white-space: nowrap;
   }
}

// makes the dropdown cover the full width of the widget
// 925px is the width that is used in the Kambi client for similar buttons
@media (max-width: 925px) {
   .general {
      position: initial;
   }

   .dropDown {
      left: 0;
      right: 0;
      /* vw means viewport width (iframe width)
      we need to use it here because of the parent div position: relative
      vw support is pretty good: http://caniuse.com/#feat=viewport-units
      but we have a fallback just in case */
      width: 280px; // fallback in case the browser doesn't support calc() or vw
      width: 100vw;
      width: calc(100vw - 8px);
   }
}

@media (min-width: 925px) {
   // yeah I know !important is bad, shut up
   // it was either this or check for the page size in JS (and having a page resize event listener)
   // these override dropdownMobileStyling in the render() of the component
   .leftAlign {
      right: auto !important;
      left: 0 !important;
   }

   .rightAlign {
      right: 0 !important;
      left: auto !important;
   }

   .topAlign {
      top: 0 !important;
      bottom: auto !important;
   }

   .bottomAlign {
      top: auto !important;
      bottom: 0 !important;
   }
}
