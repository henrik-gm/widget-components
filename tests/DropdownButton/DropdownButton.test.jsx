/* eslint-env jest */
import React, { Children } from 'react';
import DropdownButton from '../../src/DropdownButton/DropdownButton';
import ReactTestRenderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

describe('DropdownButton DOM rendering', () => {

   it('renders correctly with default props', () => {
      const options = ['option1', 'option2'];

      const tree = ReactTestRenderer.create(
         <DropdownButton options={options} onChange={() => {}} />
      ).toJSON();

      expect(tree).toMatchSnapshot();
   });

   it('renders correctly with changed alignment', () => {
      const options = ['option1', 'option2'];

      const tree = ReactTestRenderer.create(
         <DropdownButton
            options={options}
            onChange={() => {}}
            horizontalAlignment='left'
            verticalAlignment='bottom'
         />
      );

      expect(tree.toJSON()).toMatchSnapshot();
   });

});

describe('DropdownButton behaviour', () => {

   it('expands and collapses correctly', () => {
      const options = ['option1', 'option2'],
         clickEvMock = { stopPropagation: jest.fn() };

      const wrapper = mount(
         <DropdownButton
            options={options}
            onChange={() => {}}
         />
      );

      wrapper.find('button').simulate('click', clickEvMock);

      expect(clickEvMock.stopPropagation).toHaveBeenCalledTimes(1);

      window.document.documentElement.dispatchEvent(new Event('click'));
   });

   it('handles switching option correctly', () => {
      const options = ['option1', 'option2'],
         onChangeMock = jest.fn(),
         clickEvMock = { stopPropagation: jest.fn() };

      const tmpClientHeight = document.documentElement.clientHeight,
         tmpInnerHeight = window.innerHeight;

      document.documentElement.clientHeight = 100;
      window.innerHeight = 0;

      const wrapper = mount(
         <DropdownButton
            options={options}
            onChange={onChangeMock}
            verticalAlignment='bottom'
         />
      );

      wrapper.find('button').simulate('click', clickEvMock);

      expect(clickEvMock.stopPropagation).toHaveBeenCalledTimes(1);
      expect(onChangeMock).not.toHaveBeenCalled();

      const eventMock = new Event('click');
      Object.defineProperty(
         eventMock,
         'target',
         {
            value: wrapper.find('li[data-kw-dropdown-button-index=1]').node,
            enumerable: true
         }
      );

      window.document.documentElement.dispatchEvent(eventMock);

      expect(onChangeMock).toHaveBeenCalledTimes(1);

      wrapper.find('button').simulate('click', clickEvMock);

      expect(clickEvMock.stopPropagation).toHaveBeenCalledTimes(2);

      window.document.documentElement.dispatchEvent(eventMock);

      expect(onChangeMock).toHaveBeenCalledTimes(1); // not called again

      document.documentElement.clientHeight = tmpClientHeight;
      window.innerHeight = tmpInnerHeight;
   });

});
