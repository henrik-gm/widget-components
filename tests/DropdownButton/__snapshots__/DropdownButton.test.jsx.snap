exports[`DropdownButton DOM rendering renders correctly with changed alignment 1`] = `
<div
  className="general">
  <button
    className="KambiWidget-card-support-text-color button"
    onClick={[Function]}>
    option1
    <i />
  </button>
</div>
`;

exports[`DropdownButton DOM rendering renders correctly with default props 1`] = `
<div
  className="general">
  <button
    className="KambiWidget-card-support-text-color button"
    onClick={[Function]}>
    option1
    <i />
  </button>
</div>
`;
